core/src/main/java/com/datastax/oss/driver/internal/core/metadata/token/NetworkTopologyReplicationStrategy.java

```diff
@@ -56,21 +56,31 @@
   }
 
   @Override
-  public SetMultimap<Token, Node> computeReplicasByToken(
+  public Map<Token, Set<Node>> computeReplicasByToken(
       Map<Token, Node> tokenToPrimary, List<Token> ring) {
 
-    // This is essentially a copy of org.apache.cassandra.locator.NetworkTopologyStrategy
-    ImmutableSetMultimap.Builder<Token, Node> result = ImmutableSetMultimap.builder();
+    // The implementation of this method was adapted from
+    // org.apache.cassandra.locator.NetworkTopologyStrategy
+
+    ImmutableMap.Builder<Token, Set<Node>> result = ImmutableMap.builder();
     Map<String, Set<String>> racks = getRacksInDcs(tokenToPrimary.values());
     Map<String, Integer> dcNodeCount = Maps.newHashMapWithExpectedSize(replicationFactors.size());
     Set<String> warnedDcs = Sets.newHashSetWithExpectedSize(replicationFactors.size());
+    // To optimize memory usage, ensure that identical replica sets will be represented by a single
+    // instance. Order matters so the key must be a list.
+    Map<List<Node>, Set<Node>> canonicalReplicaSets = new HashMap<>();
+
     // find maximum number of nodes in each DC
     for (Node node : Sets.newHashSet(tokenToPrimary.values())) {
       String dc = node.getDatacenter();
       dcNodeCount.putIfAbsent(dc, 0);
       dcNodeCount.put(dc, dcNodeCount.get(dc) + 1);
     }
+    // Preserve order - primary replica will be first
```

Comment:
Is this list sorted somehow? If so, couldn't we implement `addIfAbsent` using `Collections.binarySearch()`?

Reply:
Also if the operations performed on this list are `add(Node)` and `contains(Node)` mostly (+ `clear()`), maybe a `LinkedList` would perform better since add is O(1).

Reply:
No, the list is not sorted, it's just the primary first and then the replicas in the order they were found while walking the ring.
I don't think `LinkedList` will perform better, it [almost never does](https://errorprone.info/bugpattern/JdkObsolete), and since we're reusing the array list it's only allocated twice, it should be pretty cheap.